### PR TITLE
Remove setter decorator for Parent.add_child method

### PR DESCRIPTION
**Pull request by mooksel**
Method `add_child` does not behave like a `setter` for a `children` property.
Actually, `children` property is read-only.

### CODE REVIEW

**Comment by fantix** [comment]
> Method add_child does not behave like a setter for a children property.

You are right about the behavior - that's why it is named `add_child` instead of `children`. Unfortunately the `setter` decorator cannot be removed because it is used by the loader:

```python
parents = await query.gino.load(
    Parent.distinct(Parent.id).load(add_child=Child.distinct(Child.id))).all()
```

**Comment by mooksel** [comment]
> > Method add_child does not behave like a setter for a children property.
> 
> You are right about the behavior - that's why it is named `add_child` instead of `children`. Unfortunately the `setter` decorator cannot be removed because it is used by the loader:
> 
> ```python
> parents = await query.gino.load(
>     Parent.distinct(Parent.id).load(add_child=Child.distinct(Child.id))).all()
> ```

Loader uses `Parent.add_child` method directly, but not uses `Parent.children` setter. If it used setter it would be look like:
```python
parents = await query.gino.load(
    Parent.distinct(Parent.id).load(children=Child.distinct(Child.id))).all()
```
So, if we remove `@children.setter` decorator nothing will change. Loader will still use `Parent.add_child` method dicretly.

One other point is that this decorator breaks static type checking. Because type checkers like `mypy` expect that `setter` will accept argument with same type as appropriate `getter` returns.

**Comment by o-podolyan** [comment]
@fantix, I see it uses `setattr` for each loaded element instead of set the whole list.
Is there a reason why this was realised in such way?

**Comment by mooksel** [comment]
@fantix, I see it uses setattr for each loaded element instead of set the whole list.
Is there a reason why this was realised in such way?

**Comment by fantix** [comment]
> @fantix, I see it uses setattr for each loaded element instead of set the whole list.
> Is there a reason why this was realised in such way?
> https://github.com/python-gino/gino/blob/ef1d9826f29540ef4a2756dfa05d046386a96689/src/gino/loader.py#L121

The way loaders work is row-oriented. Consider a result set from a joint query like this:

```sql
SELECT * FROM books LEFT JOIN authors ON books.author_id = authors.id;
```

| books.id  | books.name | authors.id | authors.name |
| ------------- | ------------- | ------------- | ------------- |
| 1  | The Mythical Man-Month  |  1  |  Frederick Phillips |
| 2  | The Design of Design  | 1  | Frederick Phillips  |
| 3  | The Art of Computer Programming  | 2  |  Donald Ervin Knuth |
| 4  | The Algorithm Design Manual  | 3  | Steven Skiena  |

The loader looks like this:

```python
Book.load(author=Author)
```

For each row, a `Book` instance is created, with its attribute `author` set to a new `Author` instance (pseudocode below):

```python
for row in result:
    book = Book(id=row['books.id'], name=row['books.name'])
    author = Author(id=row['authors.id'], name=row['authors.name'])
    book.author = author
    yield book
```

Just by the way, distinct works this way:

```python
Book.load(author=Author.distinct(Author.id))
```

Works like this pseudocode:

```python
authors = {}
for row in result:
    book = Book(id=row['books.id'], name=row['books.name'])
    if row['authors.id'] in authors:
        author = authors[row['authors.id']]
    else:
        author = Author(id=row['authors.id'], name=row['authors.name'])
        authors[row['authors.id']] = author
    book.author = author
    yield book
```

Reversely, if we want to load all authors with their books, we could do:

```python
Author.distinct(Author.id).load(add_book=Book)
```

This translates to:

```python
authors = {}
rv = []
for row in result:
    if row['authors.id'] in authors:
        author = authors[row['authors.id']]
    else:
        author = Author(id=row['authors.id'], name=row['authors.name'])
        authors[row['authors.id']] = author
        rv.append(author)
    book = Book(id=row['books.id'], name=row['books.name'])
    author.add_book = book
yield from rv
```

I think we can add some logic here to make `add_book` adaptive, for example:

```python
    if isinstance(Author.add_book, property):
        author.add_book = book
    else:
        author.add_book(book)
```

So that we could remove the "misleading" setter property decorator on `add_book`. Please feel free to update this PR to include that.

Full testing code below:

```python
from gino import Gino

db = Gino()


class Book(db.Model):
    __tablename__ = "books"

    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.Unicode)
    author_id = db.Column(db.ForeignKey("authors.id"))


class Author(db.Model):
    __tablename__ = "authors"

    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.Unicode)

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._books = set()

    @property
    def books(self):
        return self._books

    @books.setter
    def add_book(self, book):
        self._books.add(book)


async def main():
    await db.set_bind("postgresql://localhost/gino")
    await db.gino.create_all()
    a1 = await Author.create(name="Frederick Phillips")
    await Book.create(name="The Mythical Man-Month", author_id=a1.id)
    await Book.create(name="The Design of Design", author_id=a1.id)
    a2 = await Author.create(name="Donald Ervin Knuth")
    await Book.create(name="The Art of Computer Programming", author_id=a2.id)
    a3 = await Author.create(name="Steven Skiena")
    await Book.create(name="The Algorithm Design Manual", author_id=a3.id)

    authors = (
        await Book.outerjoin(Author)
        .select()
        .gino.load(Author.distinct(Author.id).load(add_book=Book))
        .all()
    )
    for author in authors:
        print(author.name)
        for book in author.books:
            print('\t', book.name)

    await db.gino.drop_all()


if __name__ == "__main__":
    import asyncio

    asyncio.run(main())
```

**Comment by fantix** [comment]
The proposed change is:

```diff
diff --git a/src/gino/loader.py b/src/gino/loader.py
index 5062615..cbbeef2 100644
--- a/src/gino/loader.py
+++ b/src/gino/loader.py
@@ -118,7 +118,10 @@ class ModelLoader(Loader):
             for key, value in self.extras.items():
                 value, distinct_ = value.do_load(row, context)
                 if distinct_ is not None:
-                    setattr(rv, key, value)
+                    if isinstance(getattr(self.model, key), property):
+                        setattr(rv, key, value)
+                    else:
+                        getattr(rv, key)(value)
             return rv, distinct

     def get_columns(self):
```

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/loader.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=4921464&bid=16285955)

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/loader.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=4921464&bid=16285955)

File: src/gino/loader.py [hunk 1]
```diff
@@ -117,7 +117,13 @@ def do_load(self, row, context):
         else:
             for key, value in self.extras.items():
                 value, distinct_ = value.do_load(row, context)
-                if distinct_ is not None:
+                if distinct_ is None:
+                    continue
+                if hasattr(self.model, key) and not isinstance(
+                    getattr(self.model, key), property
+                ):
```
Comment (fantix) [inline src/gino/loader.py]: ~(trivial) can be simplified as:~

```python
if not isinstance(getattr(self.model, key, None), (property, type(None))):
```
---
I'm so sorry - while writing above comment, I thought about another scenario that we may break here:

```python
class Book(db.Model):
    __tablename__ = "authors"

    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.Unicode)
    author = None
```

And:

```python
Book.load(author=Author)
```

We may need to check for function type?

```python
if isinstance(getattr(self.model, key, None), types.FunctionType):
```

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/loader.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=4921464&bid=16285955)

File: src/gino/loader.py (hunk 1 above)
Comment (mooksel) [reply src/gino/loader.py]: Yes, agree.
Added check for `types.FunctionType`

**Comment by mooksel** [comment]
> Thanks for the PR, I love the tests! By the way, could you please rebase to the latest master where I fixed the failing test? Appreciate it!

Done

**Comment by fantix** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/gino/loader.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=4921464&bid=16285955)

**Pull request merged**